File: frontend/src/types/feedback.ts
/**
 * Feedback type definitions for Dream LMS Teacher Feedback.
 * Story 6.4: Teacher Feedback on Assignments
 * Story 6.5: Feedback Enhancements (Badges & Emoji Reactions)
 */

/**
 * Badge information type (Story 6.5)
 */
export interface BadgeInfo {
  slug: string
  label: string
  icon: string
}

/**
 * Emoji reaction information type (Story 6.5)
 */
export interface EmojiInfo {
  slug: string
  emoji: string
}

/**
 * Predefined badges matching backend constants (Story 6.5, AC: 2)
 */
export const PREDEFINED_BADGES: BadgeInfo[] = [
  { slug: "perfect_score", label: "Perfect Score", icon: "💯" },
  { slug: "great_improvement", label: "Great Improvement", icon: "📈" },
  { slug: "creative_thinking", label: "Creative Thinking", icon: "💡" },
  { slug: "hard_worker", label: "Hard Worker", icon: "💪" },
  { slug: "fast_learner", label: "Fast Learner", icon: "⚡" },
  { slug: "needs_review", label: "Needs Review", icon: "📚" },
]

/**
 * Available emoji reactions matching backend constants (Story 6.5, AC: 5)
 */
export const AVAILABLE_EMOJI_REACTIONS: EmojiInfo[] = [
  { slug: "thumbs_up", emoji: "👍" },
  { slug: "heart", emoji: "❤️" },
  { slug: "star", emoji: "⭐" },
  { slug: "party", emoji: "🎉" },
  { slug: "fire", emoji: "🔥" },
  { slug: "hundred", emoji: "💯" },
]

/**
 * Badge labels lookup for display
 */
export const BADGE_LABELS: Record<string, string> = Object.fromEntries(
  PREDEFINED_BADGES.map((b) => [b.slug, b.label])
)

/**
 * Badge icons lookup for display
 */
export const BADGE_ICONS: Record<string, string> = Object.fromEntries(
  PREDEFINED_BADGES.map((b) => [b.slug, b.icon])
)

/**
 * Emoji display lookup
 */
export const EMOJI_DISPLAY: Record<string, string> = Object.fromEntries(
  AVAILABLE_EMOJI_REACTIONS.map((e) => [e.slug, e.emoji])
)

/**
 * Request payload for creating feedback
 */
export interface FeedbackCreate {
  feedback_text: string
  is_draft?: boolean
  badges?: string[] // Story 6.5: Badge slugs to award
  emoji_reaction?: string | null // Story 6.5: Single emoji reaction slug
}

/**
 * Request payload for updating feedback
 */
export interface FeedbackUpdate {
  feedback_text?: string
  is_draft?: boolean
  badges?: string[] // Story 6.5: Badge slugs to update
  emoji_reaction?: string | null // Story 6.5: Emoji reaction slug to update
}

/**
 * Student badge counts response (Story 6.5, AC: 9, 14)
 */
export interface StudentBadgeCountsResponse {
  badge_counts: Record<string, number>
  total: number
  this_month: Record<string, number>
  this_month_total: number
}

/**
 * Full feedback response (for teachers)
 */
export interface FeedbackPublic {
  id: string
  assignment_student_id: string
  teacher_id: string
  feedback_text: string | null
  badges: string[]
  emoji_reactions: string[]
  is_draft: boolean
  created_at: string // ISO 8601 datetime string
  updated_at: string // ISO 8601 datetime string
  assignment_id: string
  assignment_name: string
  student_id: string
  student_name: string
  student_user_id: string
  teacher_name: string
  teacher_user_id: string
  score: number | null
}

/**
 * Limited feedback view for students
 */
export interface FeedbackStudentView {
  id: string
  feedback_text: string | null
  badges: string[]
  emoji_reactions: string[]
  created_at: string // ISO 8601 datetime string
  updated_at: string // ISO 8601 datetime string
  teacher_name: string
  teacher_user_id: string
  assignment_name: string
  assignment_id: string
}

/**
 * Union type for feedback responses (API can return either)
 */
export type FeedbackResponse = FeedbackPublic | FeedbackStudentView

/**
 * Type guard to check if feedback is the full public version
 */
export function isFeedbackPublic(
  feedback: FeedbackResponse
): feedback is FeedbackPublic {
  return "is_draft" in feedback && "student_id" in feedback
}

/**
 * Type guard to check if feedback is the student view
 */
export function isFeedbackStudentView(
  feedback: FeedbackResponse
): feedback is FeedbackStudentView {
  return !("is_draft" in feedback)
}

File: backend/app/schemas/feedback.py
"""Pydantic schemas for Feedback API requests/responses."""

import uuid
from datetime import datetime
from typing import TypedDict

from pydantic import BaseModel, ConfigDict, Field, field_validator


class BadgeInfo(TypedDict):
    """Type definition for badge information."""

    slug: str
    label: str
    icon: str


class EmojiInfo(TypedDict):
    """Type definition for emoji reaction information."""

    slug: str
    emoji: str


class FeedbackCreate(BaseModel):
    """Schema for creating feedback on a student assignment."""

    feedback_text: str = Field(..., min_length=1, max_length=1000)
    is_draft: bool = Field(default=False)
    badges: list[str] | None = Field(default=None, max_length=6)
    emoji_reaction: str | None = Field(default=None)

    @field_validator("feedback_text")
    @classmethod
    def validate_feedback_not_empty(cls, v: str) -> str:
        """Validate feedback text is not empty or whitespace."""
        if not v.strip():
            raise ValueError("Feedback text cannot be empty")
        return v

    @field_validator("badges")
    @classmethod
    def validate_badges(cls, v: list[str] | None) -> list[str] | None:
        """Validate badges are from predefined list."""
        if v is None:
            return v
        # Import here to avoid circular imports
        from app.core.feedback_constants import MAX_BADGES_PER_FEEDBACK, VALID_BADGE_SLUGS

        if len(v) > MAX_BADGES_PER_FEEDBACK:
            raise ValueError(f"Maximum {MAX_BADGES_PER_FEEDBACK} badges allowed")
        invalid_badges = [b for b in v if b not in VALID_BADGE_SLUGS]
        if invalid_badges:
            raise ValueError(f"Invalid badge(s): {', '.join(invalid_badges)}")
        return v

    @field_validator("emoji_reaction")
    @classmethod
    def validate_emoji_reaction(cls, v: str | None) -> str | None:
        """Validate emoji reaction is from available list."""
        if v is None:
            return v
        from app.core.feedback_constants import VALID_EMOJI_SLUGS

        if v not in VALID_EMOJI_SLUGS:
            raise ValueError(f"Invalid emoji reaction: {v}")
        return v


class FeedbackUpdate(BaseModel):
    """Schema for updating feedback."""

    feedback_text: str | None = Field(None, min_length=1, max_length=1000)
    is_draft: bool | None = None
    badges: list[str] | None = Field(default=None, max_length=6)
    emoji_reaction: str | None = Field(default=None)

    @field_validator("feedback_text")
    @classmethod
    def validate_feedback_not_empty(cls, v: str | None) -> str | None:
        """Validate feedback text is not empty or whitespace if provided."""
        if v is not None and not v.strip():
            raise ValueError("Feedback text cannot be empty")
        return v

    @field_validator("badges")
    @classmethod
    def validate_badges(cls, v: list[str] | None) -> list[str] | None:
        """Validate badges are from predefined list."""
        if v is None:
            return v
        from app.core.feedback_constants import MAX_BADGES_PER_FEEDBACK, VALID_BADGE_SLUGS

        if len(v) > MAX_BADGES_PER_FEEDBACK:
            raise ValueError(f"Maximum {MAX_BADGES_PER_FEEDBACK} badges allowed")
        invalid_badges = [b for b in v if b not in VALID_BADGE_SLUGS]
        if invalid_badges:
            raise ValueError(f"Invalid badge(s): {', '.join(invalid_badges)}")
        return v

    @field_validator("emoji_reaction")
    @classmethod
    def validate_emoji_reaction(cls, v: str | None) -> str | None:
        """Validate emoji reaction is from available list."""
        if v is None:
            return v
        from app.core.feedback_constants import VALID_EMOJI_SLUGS

        if v not in VALID_EMOJI_SLUGS:
            raise ValueError(f"Invalid emoji reaction: {v}")
        return v


class FeedbackPublic(BaseModel):
    """Schema for feedback API response with related entity names."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    assignment_student_id: uuid.UUID
    teacher_id: uuid.UUID
    feedback_text: str | None
    badges: list[str]
    emoji_reactions: list[str]
    is_draft: bool
    created_at: datetime
    updated_at: datetime
    # Related entity info
    assignment_id: uuid.UUID
    assignment_name: str
    student_id: uuid.UUID
    student_name: str
    student_user_id: uuid.UUID
    teacher_name: str
    teacher_user_id: uuid.UUID
    score: float | None


class FeedbackStudentView(BaseModel):
    """Schema for feedback as viewed by a student (hides draft status)."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    feedback_text: str | None
    badges: list[str]
    emoji_reactions: list[str]
    created_at: datetime
    updated_at: datetime
    teacher_name: str
    teacher_user_id: uuid.UUID
    assignment_name: str
    assignment_id: uuid.UUID


class FeedbackOptionsResponse(BaseModel):
    """Response schema for available feedback options (badges and emoji)."""

    badges: list[BadgeInfo]
    emoji_reactions: list[EmojiInfo]


class StudentBadgeCountsResponse(BaseModel):
    """Response schema for student badge counts (Story 6.5, AC: 9, 14)."""

    badge_counts: dict[str, int]
    total: int
    this_month: dict[str, int]
    this_month_total: int
